### PR TITLE
scheduler(cdc): add ProcessorEpoch

### LINKED ISSUE #4769: [TiCDC] (unreleased feature) High inter-node latency causes ErrProcessorDuplicateOperations
- When the new scheduler is enabled, high inter-node latency can cause ErrProcessorDuplicateOperations.
- Scenario: 100K tables with 8 TiCDC nodes. Injects a latency of `2000ms` jitter `10ms` between TiCDC nodes. Then restarts all TiCDC nodes all at once.

```
[
  {
    "id": "test-1",
    "summary": {
      "state": "error",
      "tso": 431564666934132811,
      "checkpoint": "2022-03-03 14:22:34.894",
      "error": {
        "addr": "new-sched-test-cdc-3.liuzixiong.tispace:8300",
        "code": "CDC:ErrProcessorDuplicateOperations",
        "message": "[CDC:ErrProcessorDuplicateOperations]table processor duplicate operation, table-id: 26194"
      }
    }
  }
]
```
- Root cause: Processor may send `Sync` to Owner to tell the owner about the processor's internal state. However, the internal state can be updated by a in-flight owner's dispatch command, making the owner's knowledge of the processor' state inconsistent.
- Severity: (1) This problem rarely occurs if the latency is low (sub 100ms) (2) This problem pauses the changefeed, but the changefeed can retry and will eventually succeed if the latency recovers to normal. (3) This problem will not cause data corruption.

### PR DESCRIPTION
**Pull request by liuzix**
### What problem does this PR solve?

Issue Number: close #4769

### What is changed and how it works?
- Added `epoch` in `Sync` and `DispatchTable` messages, so that outdated dispatches will be ignored by the processor.

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test

Restarts all TiCDC nodes at once with a latency of 2000ms injected between each pair of them. The changefeed did not pause and report error, but recovered quickly.
image

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch

### Release note 

```release-note
Fix ErrProcessorDuplicateOperations when new scheduler is enabled (disabled by default)
```

### CODE REVIEW

File: cdc/scheduler/agent.go [hunk 1]
```diff
@@ -272,6 +289,7 @@ func (a *BaseAgent) processOperations(ctx context.Context) error {
 	for tableID, op := range a.tableOperations {
 		switch op.status {
 		case operationReceived:
+			a.logger.Info("Agent start processing operation", zap.Any("op", op))
```
Comment (amyangfei) [inline cdc/scheduler/agent.go:292]: Is it acceptable that the operation related log is O(#table) scale.

File: cdc/scheduler/agent.go [hunk 2]
```diff
@@ -456,3 +478,23 @@ func (a *BaseAgent) currentOwner() model.CaptureID {
 
 	return a.ownerInfo.OwnerCaptureID
 }
+
+func (a *BaseAgent) resetEpoch() {
+	a.epochMu.Lock()
+	defer a.epochMu.Unlock()
+
+	a.epoch = uuid.New().String()
```
Comment (amyangfei) [inline cdc/scheduler/agent.go:486]: We only need the unique constraint of this epoch, and don't need the serialization guarantee. Should we add comment about it.

File: cdc/scheduler/agent.go (hunk 1 above)
Comment (liuzix) [reply cdc/scheduler/agent.go:292]: Without this log, it would be difficult to trace scheduling problems.

File: cdc/scheduler/agent.go (hunk 2 above)
Comment (liuzix) [reply cdc/scheduler/agent.go:486]: Sure.

**Review by amyangfei — APPROVED** [review]

**Review by 3AceShowHand — APPROVED** [review]

**Pull request merged**